### PR TITLE
make config_json available to providers

### PR DESCRIPTION
**Pull request by stevelr**
# Feature or Problem
The `config_json` field from host_data was not accessible to providers through current apis. This change makes it available.

# Consumer Impact
Does not affect existing code

### CODE REVIEW

File: rpc-rs/src/provider.rs [hunk 1]
```diff
@@ -199,6 +197,24 @@ impl HostBridge {
...
+    /// // Example: deserialize to a hashmap
+    /// let config = host_bridge.config_json::<HashMap<String,String>>().unwrap();
+    /// # }
+    /// ```    
+    pub fn config_json<T: DeserializeOwned>(&self) -> Option<RpcResult<T>> {
+        self.host_data.config_json.as_ref().map(|json| {
+            serde_json::from_str::<T>(json.as_str())
+                .map_err(|e| RpcError::Deser(format!("deserializing config_json: {e}")))
+        })
+    }
```
Comment (brooksmtownsend) [inline rpc-rs/src/provider.rs:217]: The return type of `RpcResult` feels weird here, I know it's to specify the deserialization issue but I think it would be a little more ergonomic to have an Option<Result<T>> since this isn't an RPC operation

Comment (stevelr) [reply rpc-rs/src/provider.rs:217]: The goal was to reduce boilerplate. This is most likely to be used inside a function that returns RpcResult (when a provider needs to get configuration in response to a linkdef, or else in response to an rpc), so any caller would have to the same `map_err`.

File: rpc-rs/src/provider.rs [hunk 2]
```diff
@@ -199,6 +197,24 @@ impl HostBridge {
     pub fn lattice_prefix(&self) -> &str {
         &self.host_data.lattice_rpc_prefix
     }
+
+    /// Deserializes HostData config_json into T
+    /// (if you just need generic json, use Returns the config_json,
```
Comment (connorsmith256) [inline rpc-rs/src/provider.rs:202]: Nit: looks like bad text replacement on the end of this line

File: rpc-rs/src/provider.rs (hunk 1 above)
Comment (connorsmith256) [reply rpc-rs/src/provider.rs:217]: To keep it simple and reduce boilerplate, why not return `Option<String>`, i.e. just `self.host_data.config_json`?

The config JSON should only be parsed once. In order for wasmbus-rpc to handle that, we would need to add a strong type specified by a provider, which I assume would be a significant refactor/breaking change, and that's why you made this function generic on `T`(?) In a world where the provider _can_ specify its config type at compile-time, having this function do the parsing makes sense. But in the current state, I don't think being generic on `T` helps. The caller will still need to specify the type, so letting the caller do the parsing and error handling (once) is fine.

As a concrete example, consider a provider author wanting to check config JSON during an RPC call. With this change, they can call `config_json` without being aware that they've added a JSON parse to every invocation. Forcing the caller to do the parsing makes the cost of parsing explicit

**Review by connorsmith256 — CHANGES_REQUESTED** [review]

File: rpc-rs/src/provider.rs (hunk 1 above)
Comment (stevelr) [reply rpc-rs/src/provider.rs]: great point. I'm convinced.

**Review by connorsmith256 — APPROVED** [review]

File: rpc-rs/src/provider.rs (hunk 1 above)
Comment (thomastaylor312) [reply rpc-rs/src/provider.rs]: I'm gonna push back a bit. I think we should have _both_ versions. So have a `config_json_str() -> Option<&str>` (or `config_json_raw() -> Option<&[u8]>`) and a `config_json<T: DeserializedOwned>() -> Result<Option<T>>`. There are _many_ times someone might want to parse and having something that does it for you is super helpful. It is more about future proofing and making the library easy to use (less boilerplate if you are going to deserialize), so I think there should be both.

Don't have a super strong opinion that would make me block on having both, but I think having both of these are probably a good idea

Comment (connorsmith256) [reply rpc-rs/src/provider.rs]: To push back on the pushback, I maintain that in its current state, returning an `Option<RpcResult<T>>` is a footgun that violates the spirit of Rust 😉 

Once we can infer `T`, then this becomes much better than returning the raw JSON. _Until then_, it's worse.

There's a third option that would resolve the downside I'm concerned with, which would be lazily deserializing once and caching the value. Then a consumer can call `config_json` all they want during an RPC call and would only pay the parsing penalty once. I would also be okay with eagerly deserializing during init to avoid the first-time penalty.

(But I think this is fine as-is)

TL;DR it is critical the door to the bike shed be painted periwinkle.

Comment (thomastaylor312) [reply rpc-rs/src/provider.rs]: I don't follow your argument, but I also don't care enough to bikeshed here. We definitely should return a borrowed `&str` here though. Because even if a caller is going to deserialize, most functions take borrowed data anyway
